enclave-runner/src/usercalls/mod.rs

```diff
@@ -333,36 +380,246 @@ impl EnclaveState {
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    fn syscall_loop(io_queue_receive: mpsc::Receiver<UsercallSendData>, enclave: Arc<EnclaveState>) -> StdResult<(u64, u64), EnclaveAbort<EnclavePanic>> {
+        let main_result;
+        'outer: loop {
+            let maybe_block_recv = |block| if block {
+                io_queue_receive.recv().ok()
+            } else {
+                io_queue_receive.try_recv().ok()
+            };
+
+            'inner: while let Some((coresult, mut state, mode, buf)) = maybe_block_recv(true) {
+                let enclave = enclave.clone();
+                match coresult {
+                    CoResult::Return((tcs,v1, v2)) => {
+                        match mode {
+                            EnclaveEntry::Library {p1: _, p2: _, p3: _, p4: _, p5: _} => {
+                                let lib = enclave.kind.as_library().unwrap().thread_sender.lock().unwrap();
+                                if lib.send(StoppedTcs { tcs, event_queue: state.event_queue }).is_err() {
+                                    panic!("Work sender couldn't send data to receiver")
+                                }
+                                main_result = Ok((v1, v2));
+                                break 'outer;
+                            },
+                            EnclaveEntry::ExecutableMain => {
+                                main_result = Err(EnclaveAbort::MainReturned);
+                                break 'outer;
+                            },
+                            EnclaveEntry::ExecutableNonMain => {
+                                assert_eq!(
+                                    (v1, v2),
+                                    (0, 0),
+                                    "Expected enclave thread entrypoint to return zero"
+                                );
+                                let cmd = enclave.kind.as_command().unwrap();
+                                let mut cmddata = cmd.data.lock().unwrap();
+                                cmddata.running_secondary_threads -= 1;
+                                if cmddata.running_secondary_threads == 0 {
+                                    cmd.wait_secondary_threads.notify_all();
+                                }
+                                // If the enclave is in the exit-state, threads are no
+                                // longer able to be launched
+                                if !enclave.exiting.load(Ordering::SeqCst) {
+                                    cmddata.threads_queue.push(StoppedTcs {
+                                        tcs,
+                                        event_queue: state.event_queue,
+                                    });
+                                }
+                            }
+                        }
+                    },
+                    CoResult::Yield(usercall) => {
```

Comment:
Can you put `Yield` first in the `match`? That seems to be a more logical order to read things in. Also, can you maybe merge the Yield (error case) handling with the `CoResult::Return` handling?